File: Unique Paths II/Unique Paths II.py
class Solution:
    # @param obstacleGrid, a list of lists of integers
    # @return an integer
    def uniquePathsWithObstacles(self, obstacleGrid):
        m=len(obstacleGrid)
        if not m:
            return 0
        n=len(obstacleGrid[0])
        if not n:
            return 0
        res=[0 for i in range(n+1)]
        res[1]=1
        for i in range(m):
            for j in range(1,n+1):
                if not obstacleGrid[i][j-1]:
                    res[j]=res[j-1]+res[j]
                else:
                    res[j]=0
                    
        return res[n]

File: Unique Paths II/Unique Paths II.java
public class Solution {
    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        int m=obstacleGrid.length;
        if(m==0) return 0;
        int n=obstacleGrid[0].length;
        if(n==0) return 0;
        int[] res=new int[n+1];
        res[1]=1;
        for(int i=0;i<m;i++){
            for(int j=1;j<=n;j++){
                if(obstacleGrid[i][j-1]==0){
                    res[j]=res[j-1]+res[j];
                }else{
                    res[j]=0;
                }
            }
        }
        return res[n];
    }
}
